Add FAQ about StockAvailable and StockMovement (#964)
PrestaShop/docs: src/content/1.7/faq/stock.md

## StockAvailable

**Q:** What does the StockAvailable contain exactly?

**A:** Each `StockAvailable` is associated to a `Product` or a `Combination`, and if you are using the multi shop feature you can also have different stock for each Shop or Shop group. Here are the details of each field:

| Field               | Description |
| ------------------- | ----------- |
| `quantity`          | The remaining available quantity, this indicates which how many products can still be sold in the shop. This is what user would see on the FO if display it. It's also the quantity you can see in the product page from your BO. |
| `physical_quantity` | The actual quantity that you have in your physical stock, in your shelves. It might be different from quantity when an Order has been purchased but is not paid, or shipped yet (see `reserved_quantity`). |
| `reserved_quantity` | This indicates how many of your physical quantity is currently reserved for your pending orders, the product has been bought by a customer but it is still in your inventory because the order is not paid yet, or it is paid but the package has not been sent yet. |

> | `out_of_stock`      | This is the configuration defining how you want to handle this product in the FO when it is out of stock (available stock is too low). {{< code php >}}use PrestaShop\PrestaShop\Core\Domain\Product\Stock\ValueObject\OutOfStockType;

Author: It's not code destined to be copied so are you sure we need this PHP open tag?

However I could improve this by using `{{< highlight php >}}` it was not possible before which is why I created this code shortcut, so I guess I could remove the code shurtcut and replace it everywhere (in another PR)

> | `out_of_stock`      | This is the configuration defining how you want to handle this product in the FO when it is out of stock (available stock is too low). {{< code php >}}use PrestaShop\PrestaShop\Core\Domain\Product\Stock\ValueObject\OutOfStockType;

Reviewer: It used to be a [longstanding issue](https://github.com/alecthomas/chroma/issues/210) with the chroma highlighter, but it looks like it has been fixed recently, so you can disregard my comment 👍

| Scenario Step | quantity | physical_quantity | reserved_quantity |   |
| ------------- | -------- | ----------------- | ----------------- | - |
| `Great Product` is created in the BO with a quantity of 50 | `50` | `50` | `0` | |
| A user decides to add 3 `Great Product` into his cart      | `50` | `50` | `0` | For now the product is only in a cart, nothing is reserved anyone can still buy one of the 50 remaining products. |
| The user goes through the whole checkout process and chooses to pay by check. | `50` | `47` | `3` | The order has been created, but the payment is not accepted yet. |
| A few days later, the check is received, and the merchant changes the Order status to `Payment accepted` | `50` | `47` | `3` | The product is now paid, but it still has not left your shelves. |
| The package is prepared and delivered to a Carrier, the order status is switched to `Delivered` or `Shipped` | `47` | `47` | `0` | The product is no longer reserved, it has actually been sent to the customer. So reserved and physical quantities can be decreased. |

**Q:** So where can I see and modify my Stock?

> **A:** Although you can change the quantity directly in your product page we strongly recommend using the `Catalog > Stock` section in your BO because it is based on `StockMovement` which is more stable than direct edition.

Reviewer: ```suggestion
**A:** Although you can change the quantity directly on your product page, we strongly recommend using the `Catalog > Stock` section in your BO because it is based on `StockMovement` which is safer than a direct edition (read below).
```

## StockMovement

> **Q:** But what is a `StockMovement`? What is the difference with direct edition?

Reviewer: ```suggestion
**Q:** What is a `StockMovement`? What is the difference with direct edition?
```

> **A:** When you edit the quantity of a Product you are going to edit the StockAvailable::quantity field directly in your database. When you use a modification with a StockMovement you apply a delta to your existing stock.

Reviewer: ```suggestion
**A:** When you edit the quantity of a Product you are going to edit the `StockAvailable::quantity` field directly in your database. When you modify your stock through a StockMovement you apply a _delta_ to your existing stock.
```

> **Q:** And what is the advantage of this delta edition?

Reviewer: ```suggestion
**Q:** What is the advantage of delta edition vs direct edition?
```

> **A:** When you edit the quantity the total value is sent and saved in database, but you cannot be sure that there was no modification between the moment you open your edition page, and the moment you save the new expected value.  Imagine this scenario:

Reviewer: ```suggestion
**A:** When you edit the quantity directly, the value is saved in database, but you cannot be sure that there was no modification in stock between the moment you opened your Product page and the moment you saved the new value. Imagine this scenario:
```

> - You have a 50 `Great Product` in your stock.

Reviewer: ```suggestion
- You have 50 items of `Great Product` on stock.
```

- It is your best seller, so today you are expecting a delivery to increase your stock.
- You get 50 new products and want to udpate the stock from the Product page.
- You open the page indicating that you have 50 products left in your stock.

> - While you are checking this stock, since it is your best seller, three products have been sold and delivered so your stock is now actually 47.

Reviewer: ```suggestion
- While you are checking this stock, since it is your best seller, three items have been sold and shipped, so your stock is now actually 47, not 50.
```

- You update your expected stock from 50 to 100 and save the product.

> - Your stock now indicates 100 products, but really it should be 97 (47 + 50)

Reviewer: ```suggestion
- Your stock now shows 100 products, while it should have been 97 (50 - 3 + 50)
```

> That is why a `StockMovement` will be more stable, it will not perform the action `Change my stock to 100` but instead `Add 50 to my current stock value`, so even if the value has changed without you knowing the final value will remain correct.

Reviewer: ```suggestion
That is why a `StockMovement` is safer: it will not change your stock to 100, but add 50 to your current stock level instead. This way, even if the stock changes without your knowledge, the final value will remain correct.
```

> **Q:** Is it the only benefit of `StockMovement` ?

Reviewer: ```suggestion
**Q:** Are there any other advantages to use `StockMovement`?
```

> **A:** No, another advantage si that every operation (or every `StockMovement`) not only performs the action but is also saved in database. So you get access to a complete history of the stock evolution for the product with a few details: chat is the reason of the movement, which order or which employee performed the action, when did it happen, ... You can access this history in the `Catalog > Stock > Movements` section from your BO.

Reviewer: ```suggestion
**A:** Another advantage is that every operation (or `StockMovement`) registers a log in the database. This provides you with a complete history of the stock evolution for the product, including useful details: what was the reason for the movement, which order or which employee performed the action, when did it happen, and more. You can access this history in the `Catalog > Stock > Movements` section from your BO.
```

You will probably notice that usually ```physical_quantity = available_quantity + reserved_quantity;```
{{% /notice %}}

{{% notice note %}}

> To avoid confusion when referring to `StockAvailable::quantity` below we will use the **virtual quantity** expression.

Reviewer: ```suggestion
To avoid confusion, from now on we will refer to `StockAvailable::quantity` as **virtual quantity**.
```

| A few days later, the merchant receives the check, so they change the Order status to `Payment accepted` | `47` | `50` | `3` | The product is now paid, but the package has still not been prepared. |
| The package is prepared and delivered to a Carrier, the order status is switched to `Delivered` or `Shipped` | `47` | `47` | `0` | The product is no longer reserved, it has actually been sent to the customer. So reserved and physical quantities can be decreased. |

**Q:** Where can I see and modify my Stock?

**A:** Although you can change the quantity directly on your product page, we strongly recommend using the `Catalog > Stock` section in your BO because it is based on `StockMovement` which is safer than a direct edition ([read below why](#stockmovement)).

**Q:** Can I edit the physical or reserved quantity myself or using an ERP?

**A:** Avoid doing this! These two fields should be considered read-only and exist for performance reasons. They are automatically recomputed every time the stock is synchronized:

> - **reserved_quantity** is the sum of each quantity in pending orders:

Reviewer: You need two spaces at the end of the line for the line break to be taken into account
```suggestion
- **reserved_quantity** is the sum of each quantity in pending orders:  
```

  `reserved_quantity_per_order = SUM(order_detail.product_quantity - order_detail.product_quantity_refunded)`

> - **physical_quantity** is recalculated after the reserved quantity is updated:

Reviewer: Same here
```suggestion
- **physical_quantity** is recalculated after the reserved quantity is updated:  
```

| A customer decides to add 3 `Great Product` into their cart | `50` | `50` | `0` | For now the product is only in a cart, nothing is reserved, anyone can still buy one of the 50 remaining products. |
| The customer goes through the whole checkout process and chooses to pay by check. | `47` | `50` | `3` | The order has been created, but the payment is not accepted yet. |
| A few days later, the merchant receives the check, so they change the Order status to `Payment accepted` | `47` | `50` | `3` | The product is now paid, but it still has not left the shelves. |
| The package is prepared and delivered to a Carrier, the order status is switched to `Delivered` or `Shipped` | `47` | `47` | `0` | The product is no longer reserved, it has actually been sent to the customer. So reserved and physical quantities can be decreased. |

**Q:** Where can I see and modify my Stock?

**A:** Although you can change the quantity directly on your product page, we strongly recommend using the `Catalog > Stock` section in your BO because it is based on `StockMovement` which is safer than a direct edition ([read below](#stockmovement)).

> **Q:** Can I edit the physical or reserved quantity myself? or using an ERP?

Reviewer: ```suggestion
**Q:** Can I edit the physical or reserved quantity myself or using an ERP?
```

> **A:** NO, these two fields should be considered as read only, they made the display of stock data faster. They are automatically recomputed when the stock is synchronized:

Reviewer: ```suggestion
**A:** Avoid doing this! These two fields should be considered read-only and exist for performance reasons. They are automatically recomputed every time the stock is synchronized:
```

> - **reserved_quantity**: is the sum of each quantity in pending orders `reserved_quantity_per_order = SUM(order_detail.product_quantity - order_detail.product_quantity_refunded)`

Reviewer: ```suggestion
- **reserved_quantity** is the sum of each quantity in pending orders: 
  `reserved_quantity_per_order = SUM(order_detail.product_quantity - order_detail.product_quantity_refunded)`
```

> - **physical_quantity**: once the reserved quantity is updated based on Orders, the physical one is computed like this `physical_quantity = available_quantity + reserved_quantity`

Reviewer: ```suggestion
- **physical_quantity** is recalculated after the reserved quantity is updated: 
  `physical_quantity = available_quantity + reserved_quantity`
```

> So in the end the only important field is `StockAvailable::quantity`, the other two fields are handled by the core. If you decide to modify them yourself, you should be very careful.

Reviewer: ```suggestion
So in the end the only important field is `StockAvailable::quantity`, the other two fields are handled by the core. You should avoid modifying them.
```
